File: sdk/typescript/examples/weather-agent.ts
import { Registry, Agent, createTool, z } from '../src/index.js';

/**
 * Example: Weather Agent with anybrowse TypeScript SDK
 * 
 * Shows how to:
 * 1. Define a weather tool with Zod validation
 * 2. Register it with anybrowse
 * 3. Discover and use it from another agent
 */

// ============================================================================
// PART 1: Define a Weather Tool
// ============================================================================

const weatherTool = createTool({
  name: 'get_weather',
  description: 'Get current weather conditions for a location. Use this when you need weather data including temperature, conditions, and forecast.',
  schema: {
    location: z.string().describe('City name or coordinates'),
    units: z.enum(['celsius', 'fahrenheit']).default('celsius').describe('Temperature units'),
  },
  handler: async ({ location, units }) => {
    // Mock weather API - replace with real API call
    const mockData: Record<string, { temp: number; condition: string }> = {
      'San Francisco': { temp: 18, condition: 'Foggy' },
      'New York': { temp: 25, condition: 'Sunny' },
      'London': { temp: 15, condition: 'Rainy' },
    };

    const data = mockData[location] || { temp: 20, condition: 'Clear' };
    
    let temperature = data.temp;
    if (units === 'fahrenheit') {
      temperature = temperature * 9/5 + 32;
    }

    return {
      location,
      temperature,
      units,
      condition: data.condition,
    };
  },
});

// ============================================================================
// PART 2: Temperature Conversion Tool
// ============================================================================

const convertTool = createTool({
  name: 'convert_temperature',
  description: 'Convert temperature between celsius and fahrenheit',
  schema: {
    temperature: z.number(),
    fromUnit: z.enum(['celsius', 'fahrenheit']),
    toUnit: z.enum(['celsius', 'fahrenheit']),
  },
  handler: async ({ temperature, fromUnit, toUnit }) => {
    if (fromUnit === toUnit) {
      return { temperature, unit: toUnit };
    }

    let converted: number;
    if (fromUnit === 'celsius' && toUnit === 'fahrenheit') {
      converted = temperature * 9/5 + 32;
    } else {
      converted = (temperature - 32) * 5/9;
    }

    return { temperature: converted, unit: toUnit };
  },
});

// ============================================================================
// PART 3: Register with anybrowse
// ============================================================================

async function registerWeatherServer(): Promise<string | undefined> {
  const apiKey = process.env.ANYBROWSE_API_KEY;
  if (!apiKey) {
    console.log('Set ANYBROWSE_API_KEY to register');
    return;
  }

  const registry = new Registry({ apiKey });

  const serverId = await registry.registerServer(
    'weather-provider',
    [weatherTool, convertTool],
    'Provides current weather data and temperature conversions',
    { author: 'Your Name', license: 'MIT' }
  );

  console.log(`✅ Weather server registered: ${serverId}`);
  return serverId;
}

// ============================================================================
// PART 4: Discover and Use Tools
// ============================================================================

async function discoverAndUse(): Promise<void> {
  const registry = new Registry();

  console.log('🔍 Discovering weather tools...');
  const tools = await registry.discover('weather data', { limit: 5 });

  if (tools.length === 0) {
    console.log('No weather tools found');
    return;
  }

  console.log(`Found ${tools.length} weather tools:`);
  for (const tool of tools) {
    console.log(`  - ${tool.name}: ${tool.description}`);
  }

  // Use the first tool
  const weatherTool = tools[0];
  console.log(`\n🌤️ Using ${weatherTool.name}...`);

  try {
    const result = await weatherTool.call({
      location: 'San Francisco',
      units: 'celsius',
    });
    console.log('Result:', result);
  } catch (error) {
    console.error('Error calling tool:', error);
  }
}

// ============================================================================
// PART 5: Agent Interface
// ============================================================================

async function useAgentInterface(): Promise<void> {
  const registry = new Registry();
  const agent = new Agent(registry);

  console.log('🤖 Agent finding and using weather tool...');

  try {
    const result = await agent.useTool('get weather for San Francisco', {
      location: 'San Francisco',
      units: 'celsius',
    });
    console.log('Weather:', result);
  } catch (error) {
    console.error('Error:', error);
  }
}

// ============================================================================
// MAIN
// ============================================================================

async function main(): Promise<void> {
  console.log('='.repeat(60));
  console.log('anybrowse SDK Example: Weather Agent (TypeScript)');
  console.log('='.repeat(60));

  // Example 1: Local tool usage
  console.log('\n1️⃣ Local tool usage:');
  const result = await weatherTool.call({ location: 'San Francisco', units: 'celsius' });
  console.log(`   Weather in SF: ${result.temperature}°C, ${result.condition}`);

  // Example 2: Temperature conversion
  console.log('\n2️⃣ Temperature conversion:');
  const converted = await convertTool.call({ temperature: 20, fromUnit: 'celsius', toUnit: 'fahrenheit' });
  console.log(`   20°C = ${converted.temperature}°F`);

  // Example 3: Registration
  console.log('\n3️⃣ Registration (requires ANYBROWSE_API_KEY):');
  await registerWeatherServer();

  // Example 4: Discovery
  console.log('\n4️⃣ Discovery (requires registered servers):');
  try {
    await discoverAndUse();
  } catch (e) {
    console.log('   (Skipped - no API key or registered servers)');
  }

  console.log('\n' + '='.repeat(60));
  console.log('Example complete!');
  console.log('='.repeat(60));
}

main().catch(console.error);

File: sdk/python/examples/weather_agent.py
"""
Example: Weather Agent with anybrowse SDK

This example shows how to:
1. Define a weather tool
2. Register it with anybrowse
3. Discover and use it from another agent
"""

import os
from anybrowse import Registry, Tool, Agent, tool


# ============================================================================
# PART 1: Define and Register a Weather Tool
# ============================================================================

def create_weather_tool():
    """Create a weather tool with a local handler."""
    
    def get_weather_handler(location: str, units: str = "celsius"):
        """Mock weather API - replace with real API call."""
        # In production, call OpenWeatherMap, WeatherAPI, etc.
        mock_data = {
            "San Francisco": {"temp": 18, "condition": "Foggy"},
            "New York": {"temp": 25, "condition": "Sunny"},
            "London": {"temp": 15, "condition": "Rainy"},
        }
        
        data = mock_data.get(location, {"temp": 20, "condition": "Clear"})
        
        if units == "fahrenheit":
            data["temp"] = data["temp"] * 9/5 + 32
        
        return {
            "location": location,
            "temperature": data["temp"],
            "units": units,
            "condition": data["condition"]
        }
    
    return Tool(
        name="get_weather",
        description="Get current weather conditions for a location. Use this when you need temperature, conditions, or forecast data.",
        schema={
            "location": str,
            "units": str
        },
        handler=get_weather_handler
    )


def register_weather_server():
    """Register the weather tool with anybrowse."""
    
    api_key = os.environ.get("ANYBROWSE_API_KEY")
    if not api_key:
        print("Set ANYBROWSE_API_KEY to register")
        return
    
    registry = Registry(api_key=api_key)
    weather_tool = create_weather_tool()
    
    server_id = registry.register_server(
        name="weather-provider",
        tools=[weather_tool],
        description="Provides current weather data for any location",
        metadata={
            "author": "Your Name",
            "license": "MIT"
        }
    )
    
    print(f"✅ Weather server registered: {server_id}")
    return server_id


# ============================================================================
# PART 2: Using the Decorator Syntax
# ============================================================================

@tool(
    "Convert temperature between celsius and fahrenheit",
    temperature=float,
    from_unit=str,
    to_unit=str
)
def convert_temperature(temperature: float, from_unit: str, to_unit: str) -> dict:
    """Temperature conversion tool."""
    if from_unit == to_unit:
        return {"temperature": temperature, "unit": to_unit}
    
    if from_unit == "celsius" and to_unit == "fahrenheit":
        converted = temperature * 9/5 + 32
    elif from_unit == "fahrenheit" and to_unit == "celsius":
        converted = (temperature - 32) * 5/9
    else:
        raise ValueError(f"Unknown units: {from_unit}, {to_unit}")
    
    return {"temperature": converted, "unit": to_unit}


# ============================================================================
# PART 3: Discover and Use Tools
# ============================================================================

def discover_and_use():
    """Example of discovering and using tools."""
    
    registry = Registry()  # Uses ANYBROWSE_API_KEY from env
    
    # Discover weather tools
    print("🔍 Discovering weather tools...")
    tools = registry.discover("weather data", limit=5)
    
    if not tools:
        print("No weather tools found")
        return
    
    print(f"Found {len(tools)} weather tools:")
    for tool in tools:
        print(f"  - {tool.name}: {tool.description}")
    
    # Use the first tool
    weather_tool = tools[0]
    print(f"\n🌤️ Using {weather_tool.name}...")
    
    if weather_tool.handler:
        result = weather_tool.call(location="San Francisco", units="celsius")
        print(f"Result: {result}")
    else:
        # Remote tool - call through registry
        result = registry.call_tool(
            weather_tool.server_id,
            weather_tool.name,
            {"location": "San Francisco", "units": "celsius"}
        )
        print(f"Result: {result}")


# ============================================================================
# PART 4: Agent Interface
# ============================================================================

def use_agent_interface():
    """Example using the high-level Agent interface."""
    
    registry = Registry()
    agent = Agent(registry)
    
    # The agent automatically discovers and uses the right tool
    print("🤖 Agent finding and using weather tool...")
    
    try:
        result = agent.use_tool(
            "get weather for San Francisco",
            location="San Francisco",
            units="celsius"
        )
        print(f"Weather: {result}")
    except ValueError as e:
        print(f"Error: {e}")


# ============================================================================
# MAIN
# ============================================================================

if __name__ == "__main__":
    print("=" * 60)
    print("anybrowse SDK Example: Weather Agent")
    print("=" * 60)
    
    # Example 1: Local tool usage
    print("\n1️⃣ Local tool usage:")
    tool = create_weather_tool()
    result = tool.call(location="San Francisco", units="celsius")
    print(f"   Weather in SF: {result['temperature']}°C, {result['condition']}")
    
    # Example 2: Decorator
    print("\n2️⃣ Using decorator syntax:")
    result = convert_temperature(20, "celsius", "fahrenheit")
    print(f"   20°C = {result['temperature']}°F")
    
    # Example 3: Registration (requires API key)
    print("\n3️⃣ Registration (requires ANYBROWSE_API_KEY):")
    server_id = register_weather_server()
    
    # Example 4: Discovery (requires registered servers)
    print("\n4️⃣ Discovery (requires registered servers):")
    try:
        discover_and_use()
    except Exception as e:
        print(f"   (Skipped - no API key or registered servers)")
    
    print("\n" + "=" * 60)
    print("Example complete!")
    print("=" * 60)
